### PR TITLE
Client-side encryption

### PR DESCRIPTION
**Pull request by jyemin**
Patch build: https://evergreen.mongodb.com/version/5cfa6984e3c3313e376b9c62

### CODE REVIEW

File: driver-sync/src/main/com/mongodb/client/internal/CommandMarker.java [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+    CommandMarker(final Map<String, Object> options) {
+        String connectionString;
+
+        if (options.containsKey("mongocryptdURI")) {
+            connectionString = (String) options.get("mongocryptdURI");
+        } else if (System.getProperty("os.name").toLowerCase().contains("windows")) {
+            connectionString = "mongodb://localhost:27020";
+        } else {
+            connectionString = "mongodb://%2Ftmp%2Fmongocryptd.sock";
```
Comment (rozza) [inline driver-sync/src/main/com/mongodb/client/internal/CommandMarker.java]: Not sure what the best course of action here is.  If it fails could we add something to tell the user to install the optional dep?

File: driver-sync/src/main/com/mongodb/client/internal/CryptConnection.java [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+        rootMap.put(payload.getPayloadName(), payloadFieldNameValidator);
+        return new MappedFieldNameValidator(commandFieldNameValidator, rootMap);
+    }
+
+    private BsonWriterSettings createBsonWriterSettings() {
+        return new BsonWriterSettings();
+    }
+
+    // TODO
+    // Currently these settings will allow a split point that is potentially too large after encryption occurs. Need to find a way
```
Comment (rozza) [inline driver-sync/src/main/com/mongodb/client/internal/CryptConnection.java]: I don't recall but does the `SplittablePayloadMessage` have a min size of 1 anyway?

File: driver-sync/src/main/com/mongodb/client/vault/KeyVault.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    /**
+     * Create a data key with the given KMS provider and options.
+     *
+     * @param kmsProvider    the KMS provider
+     * @param dataKeyOptions the options for data key creation
+     * @return the identifier for the data key
+     */
+    BsonBinary createDataKey(String kmsProvider, DataKeyOptions dataKeyOptions);
+
+    // TODO should these be generic methods that can use a codec registry
```
Comment (rozza) [inline driver-sync/src/main/com/mongodb/client/vault/KeyVault.java]: Does this TODO still stand?

File: driver-core/src/test/functional/com/mongodb/client/CrudTestHelper.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+import org.bson.BsonArray;
+import org.bson.BsonDocument;
+import org.bson.BsonType;
+import org.bson.BsonValue;
+
+import static org.junit.Assert.assertEquals;
+
+public final class CrudTestHelper {
+
+    public static void replaceTypeAssertionWithActual(final BsonDocument expected, final BsonDocument actual) {
```
[the remark this answers is not in the document]
Comment (jyemin) [reply driver-core/src/test/functional/com/mongodb/client/CrudTestHelper.java]: Yeah, fun to write this one

File: driver-sync/src/main/com/mongodb/client/vault/KeyVault.java (hunk 3 above)
Comment (jyemin) [reply driver-sync/src/main/com/mongodb/client/vault/KeyVault.java]: I think it does.  BsonValue isn't the most friendly but on the other hand most users are unaware of the registry.  Could default to the default registry, or configure the registry in the settings.  What do you think?

File: driver-sync/src/main/com/mongodb/client/internal/CommandMarker.java (hunk 1 above)
Comment (jyemin) [reply driver-sync/src/main/com/mongodb/client/internal/CommandMarker.java]: Either that, make it a required dependency, or just document it.  mongo-crypt is already optional, so it's just another dependency you have to add to get this working.

File: driver-sync/src/main/com/mongodb/client/internal/CryptConnection.java (hunk 2 above)
Comment (jyemin) [reply driver-sync/src/main/com/mongodb/client/internal/CryptConnection.java]: I'm talking about size in bytes here, not size in documents.  Is that what you mean?

Comment (rozza) [reply driver-sync/src/main/com/mongodb/client/internal/CryptConnection.java]: I guess I'm not 100% sure on the todo comment

**Pull request closed without merging**